### ISSUE TITLE
Add Spotify Connect support?

### ISSUE THREAD

**Issue by idcmp** [fondberg/spotcast#154]
I've installed this component and confirmed it'll bootstrap my idle Chromecast (not listed in media_player's source_list). Sweet!

I noticed the README also mentions Spotify Connect devices too. When I put in the name of my Spotify Connect device in `device_name`, I get an error saying it couldn't find the device. I looked for some signs that it should work as I expect, but didn't find them..

I'm posting this since I'm hoping you've done more research into the two sp_ cookies I've provided to spotcast, and wonder if they can be used to do the "addUser" handshake on a Spotify Connect device directly.

### Background

Idle Spotify Connect devices have a similar problem to idle chromecasts. They need to be discovered by zeroconf and then resolved to an IP/Port/Path tuple: ` avahi-browse _spotify-connect._tcp`  gives:

```
= enp.... IPv4 Home Theater                                  _spotify-connect._tcp local
   hostname = [Home-Theater.local]
   address = [192.168.1.118]
   port = [80]
   txt = ["CPath=/spotify" "VERSION=1.0"]
```

Once discovered, they can be probed:

`curl -d action=getInfo http://192.168.1.118/spotify`

```
{
  "status": 101,
  "statusString": "OK",
  "spotifyError": 0,
  "version": "2.7.1",
  "deviceID": "4....5",
  "remoteName": "Home Theater",
  "activeUser": "",
  "accountReq": "DONTCARE",
  "deviceType": "AVR",
  "publicKey": "W....Hp",
  "brandDisplayName": "Denon",
  "modelDisplayName": "Denon ....",
  "libraryVersion": "3....",
  "resolverVersion": "0",
  "groupStatus": "NONE",
  "tokenType": "accesstoken",
  "clientID": "8f....2c",
  "productID": nn,
  "scope": "streaming",
  "availability": "",
  "voiceSupport": "NO"
}

```

I believe to safely connect to the device, you first need to boot the current user off:
`curl -d action=resetUsers http://192.168.1.118/spotify | jq -r .`

### Missing Piece

Add the user. 

The process is documented here: https://github.com/librespot-org/librespot/blob/master/docs/authentication.md (sending `action=addUser`), and there's an example of the payload to send here: 

https://github.com/badfortrains/spotcontrol/blob/869dc3fe64a91beafd767230c8413b04114babcf/discovery.go#L178

The `spotcontrol` tool seems to fail to authenticate anymore - the learning curve to figure out why looks quite steep.

I believe if this component could craft a "blob" (see the spotcontrol code), it could also connect to idle Spotify Connect devices and end a situation which has been plaguing people for _years_.

Would you be able to take a quick look?

**Comment by rchovan** [comment]
Hi, 
maybe related issue ? When I configure my Heos player, it works until spotify token is changed. Do I need to manually everyday change token, and restart HA ? 

```
Logger: homeassistant
Source: custom_components/spotcast/__init__.py:377
First occurred: December 31, 2020, 4:56:39 PM (19 occurrences)
Last logged: 11:14:45 AM

Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/config/custom_components/spotcast/__init__.py", line 373, in get_spotify_token
    self._access_token, self._token_expires = st.start_session(self.sp_dc, self.sp_key)
  File "/usr/local/lib/python3.8/site-packages/spotify_token.py", line 18, in start_session
    response = session.get("https://open.spotify.com/get_access_token?reason=transport&productType=web_player",
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 555, in get
    return self.request('GET', url, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 677, in send
    history = [resp for resp in gen]
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 677, in <listcomp>
    history = [resp for resp in gen]
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 166, in resolve_redirects
    raise TooManyRedirects('Exceeded {} redirects.'.format(self.max_redirects), response=resp)
requests.exceptions.TooManyRedirects: Exceeded 30 redirects.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/spotcast/__init__.py", line 109, in run
    return await loop.run_in_executor(executor, pfunc)
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/spotcast/__init__.py", line 184, in get_devices
    client = spotipy.Spotify(auth=get_token_instance(account).access_token)
  File "/config/custom_components/spotcast/__init__.py", line 365, in access_token
    self.ensure_token_valid()
  File "/config/custom_components/spotcast/__init__.py", line 361, in ensure_token_valid
    self.get_spotify_token()
  File "/config/custom_components/spotcast/__init__.py", line 377, in get_spotify_token
    raise HomeAssistantError("Could not get spotify token")
homeassistant.exceptions.HomeAssistantError: Could not get spotify token
```

**Comment by fondberg** [comment]
You can't use device_name for Spotify connect devices. But I agree it could be made simpler. I will see if I have time to dig into this soon

**Comment by mdbraber** [comment]
Been looking into this to and I think the Go code mentioned by @idcmp might be a good starting point, more specifically this: https://github.com/badfortrains/spotcontrol/blob/869dc3fe64a91beafd767230c8413b04114babcf/blob.go#L74

@idcmp @fondberg have you been looking into this recently by any chance?

**Comment by fondberg** [comment]
What is missing? 
The parameter is not the same though. Have a loik here: https://github.com/fondberg/spotcast/blob/master/custom_components/spotcast/services.yaml#L7

**Comment by idcmp** [comment]
I did spend a fair bit of time digging into this. There's a handshake that happens as part of Spotify Connect in which the Spotify client (app/player) exchanges some credentials which allow the output device to dynamically authenticate to the Spotify servers with/as your account. There's some authentication material that I think is actually unavailable outside the client - so it's not possible to have an idle Spotify Connect device and have it start playing music under Spotify Connect with just API calls.

There is the librespot based player - but that one takes your username and password to authenticate to Spotify servers directly - it's not like the regular Spotify Connect device (discoverable over mDNS, etc).

I don't know anyone at Spotify, but if I did, I would try to suss out if my analysis is correct. It does feel like an odd missing piece in an otherwise decent ecosystem (and people have been complaining about the omission for years).
